File: workflow/scripts/rasterize_xyz.R
library("rgrass")

gisbase <- system2("grass", args = c("--config", "path"), stdout = TRUE)
template <- terra::rast(matrix(), crs = "epsg:3067")
invisible(initGRASS(gisBase = gisbase, home = tempdir(), SG = template))

input_xyz <- snakemake@input[["raster"]]
output_field <- snakemake@wildcards[["gridcell_field"]]

field_type <- snakemake@params[["field_type"]]
resolution <- snakemake@params[["resolution"]]

if (field_type == "category") {
    grass_type <- "CELL"
    gdal_type <- "Byte"
    gdal_nodata <- 0
    geotiff_predictor <- 2
    nbits <- NULL
} else if (field_type %in% c("integer", "date")) {
    grass_type <- "CELL"
    if (substr(output_field, 1, 9) == "stemcount") {
        gdal_type <- "UInt32"
        gdal_nodata <- 2^20 - 1
        nbits <- "NBITS=20"
        geotiff_predictor <- 1
    } else {
        gdal_type <- "UInt16"
        gdal_nodata <- 2^16 - 1
        nbits <- NULL
        geotiff_predictor <- 2
    }
} else if (field_type == "real") {
    grass_type <- "FCELL"
    gdal_type <- "Float32"
    gdal_nodata <- -65504
    geotiff_predictor <- 3
    nbits <- "NBITS=16"
}

extent <- execGRASS(
    "r.in.xyz",
    flags = c("g", "s"),
    input = input_xyz,
    separator = "tab",
    skip = 1,
    intern = TRUE
)

extent <- strsplit(strsplit(extent, " ", fixed = TRUE)[[1]][1:4], "=")

extent <- setNames(
    lapply(extent, "[", i = 2),
    lapply(extent, "[", i = 1)
)

execGRASS(
    "g.region",
    parameters = c(extent, res = as.character(resolution))
)

execGRASS(
    "g.region",
    n = paste0("n+", resolution / 2),
    s = paste0("s-", resolution / 2),
    w = paste0("w-", resolution / 2),
    e = paste0("e+", resolution / 2)
)

execGRASS(
    "r.in.xyz",
    input = input_xyz,
    output = output_field,
    separator = "tab",
    skip = 1,
    type = grass_type
)

execGRASS(
    "r.out.gdal",
    flags = c("c"),
    input = output_field,
    output = snakemake@output[[1]],
    type = gdal_type,
    nodata = gdal_nodata,
    createopt = c("COMPRESS=LZW", paste0("PREDICTOR=", geotiff_predictor), nbits, "TILED=YES", "BIGTIFF=YES")
)

File: workflow/scripts/gridcell/rasterize_xyz.py
import typing

if typing.TYPE_CHECKING:
    from snakemake.script import snakemake

import subprocess
import sys

from os.path import join
from tempfile import TemporaryDirectory

sys.stdout = open(snakemake.log[0], "w")
sys.stderr = sys.stdout

sys.path.append(
    subprocess.check_output(["grass", "--config", "python_path"], text=True).strip()
)

from grass.script.core import create_project
from grass.script.setup import init as init_grass
from grass.experimental import TemporaryMapsetSession
from grass.tools import Tools

with TemporaryDirectory() as gisdb:
    create_project(path=join(gisdb, "project"), epsg=3067)

    with init_grass(gisdb, "project", "PERMANENT") as session:

        resolution = snakemake.params.resolution
        field_type = snakemake.params.field_type

        input_xyz = snakemake.input.xyz

        output_field = snakemake.wildcards.gridcell_field

        export_config = dict()

        match field_type:

            case "category" | "integer" | "date":
                grass_type = "CELL"

                match field_type:
                    case "category":
                        export_config |= {
                            "type": "Byte",
                            "nodata": 0,
                            "createopt": ["PREDICTOR=2"],
                        }

                    case "integer" if output_field[:9] == "stemcount":
                        export_config |= {
                            "type": "UInt32",
                            "nodata": 2**20 - 1,
                            "createopt": ["NBITS=20"],
                        }

                    case _:
                        export_config |= {
                            "type": "UInt16",
                            "nodata": 2**16 - 1,
                            "createopt": ["PREDICTOR=2"],
                        }

            case "real":
                grass_type = "FCELL"

                export_config |= {
                    "type": "Float32",
                    "createopt": [
                        "PREDICTOR=3",
                    ],
                }

        export_config["createopt"] += ["COMPRESS=DEFLATE", "TILED=YES"]

        tools = Tools(quiet=True)

        extent = {
            x[0]: x[1]
            for x in (
                x.split("=")
                for x in tools.r_in_xyz(
                    flags="gs", input=input_xyz, separator="tab", skip="1"
                ).space_items
            )
            if x[0] in ["n", "s", "e", "w"]
        }

        tools.g_region(res=resolution, **extent)

        tools.g_region(
            n=f"n+{resolution/2}",
            s=f"s-{resolution/2}",
            e=f"e+{resolution/2}",
            w=f"w-{resolution/2}",
        )

        tools.r_in_xyz(
            input=input_xyz,
            output=output_field,
            separator="tab",
            skip="1",
            type=grass_type,
        )

        tools.r_out_gdal(
            flags="c", input=output_field, output=snakemake.output[0], **export_config
        )
